### ISSUE TITLE
Reconsider `List.pop()` family of routines

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#16194]
While working on writing List code last week, the current pop() routines seemed problematic to me in a few ways:

* to remove the front element of the list, one has to use `pop(0)` which isn't index-neutral, or else `pop(myList.indices.low)` which is a mouthful.  Some sort of routine like `popFirst()` or `removeFirst()` seems like it would be clearer while resolving these issues.

* similarly, it seems odd to me to have a `mylist.pop()` routine that removes the last element given that lists aren't stacks, so there's no obvious end to pop from.  I think with the original list implementation, we could say that the last element is the cheapest to remove, so maybe the most obvious to pop (?), but as we consider supporting multiple list implementations under the covers, this explanation evaporates (e.g., a doubly-linked list with head/tail pointers should be equally fast).  For that reason (and for symmetry with the previous bullet), it seems like it would be better to use `popLast()` or `removeLast()`.

* depending on the choices made in the previous bullets, we may want to consider the naming of the `pop(i)` routine as well.

**Comment by gaurav-singh1998** [comment]
@bradcray: I was going through the Chapel `List` code and found this issue and I am interested to take up this task. To resolve this I just implement `popFirst()` and `popLast()`. Am I correct?

**Comment by bradcray** [comment]
Hi @gaurav-singh1998 — Thanks for your interest in running with this issue, which I appreciate.  Left to my own devices, what you're proposing is what I'd do, but since this involves a change (or at least evolution) to the current List API, this is the sort of thing that should receive some design discussion before proceeding with a specific plan to make sure that there's general consensus among the developers before proceeding.  You can read some general design-related tips at https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/ContributorInfo.rst#design where it would be reasonable to kick this design decision off either here on GitHub as an issue or as a discussion on our Discourse in the developers category.  In either case, see if you can tag those who've contributed significantly to the List module in the past to make sure that those who are most invested are in agreement.  If you have any questions about how to proceed here, please let me know.

**Comment by npadmana** [comment]
Just capturing a discussion on gitter :
*  I'd support switching to `popFirst`/`popLast` and possibly including a symmetric `pushFirst`/`pushLast` option. I'll note that the C++ STL uses `push/pop_front/back`, but I'm pretty ambivalent to the choice here.
* I'd suggest supporting both `First`/`Last` if the (possibly amortized) costs are the same. Again, from the C++ STL, `front/back` is supported for `list` while only `back` is supported for `vector`. (I haven't had a chance to look at the `List` implementation to see which one Chapel is closer to).

**Comment by jeremiah-corrado** [comment]
This issue was discussed in a recent deep dive meeting. The following is a summary of the discussion:

Concerns with the current design:
---

- There is an asymmetry with the names `pop` and `append`. The current naming comes from Python.
- As mentioned above, `pop` might be ambiguous WRT which end of the list the element is being `pop`ed from. 
    - There is also no method specifically for `pop`ing from the front of the list. This could be considered asymmetrical.
- Generally, the name `pop` could be overly reminiscent of a stack/queue like data structure. We might consider using a different name.
- The name "pop" has a strong precedent for removing an element from an **end** of a list/queue/heap/stack, so it feels like a bad name for the overload that can remove an element from any index in the list. The current design of `pop(i)` also comes from Python.

What other languages do:
---

| Chapel [list](https://chapel-lang.org/docs/modules/standard/List.html): | Python [list](https://docs.python.org/3/tutorial/datastructures.html)  | C++ [vector](https://cplusplus.com/reference/vector/vector/) | Rust [Vec](https://doc.rust-lang.org/std/vec/struct.Vec.html) | JS [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array) | Swift [Array](https://developer.apple.com/documentation/swift/array) | Go [Array/Slice](https://go.dev/blog/slices-intro) | Julia [Array](https://julia.school/julia/arrays/) | Java [ArrayList](https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html) |
| :--- | --- | ---  | --- | --- | --- | --- | --- | --- |
| `append` | `append`| `push_back` | `push` | `push`  | `append` | standalone `append` | standalone `push!` | `add` |
| `pop` |  `pop` | `pop_back` | `pop` | `pop` | `popLast` | slice & assign | standalone `pop!` | `get` & `remove` |
| `pop(i)` | `pop(i)` |  copy & `erase` | `remove` | `splice` | `remove` | slice & concat | copy & `deleteat!` | `get` & `remove` | 

Proposals:
---

The following are some proposals that were raised to address each of the above concerns:

**pop/append asymmetry**:

1. deprecate `append` and replace with `push`
     - Pros: Strong precedent from other languages. Nice symmetry.
     - Cons: could imply a queue/stack-like implementation. Could still be an end-ambiguity (which end is being pushed/popped?)
2. deprecate `append` and replace with `pushBack`, `pushLast`, or something similar
    - Pros: Precedent from C++. No ambiguity about which end of the list they refer to. Makes more sense with `[push/pop]Front` methods,  if those are added.
    - Cons: More verbose than `push`/`pop`. 
3. deprecate `pop` and replace with `removeBack`, `removeLast`, etc
    - Pros: No stack/queue connotations with this name. If `pop(i)` is renamed to `remove`, then `removeLast` would fit nicely with that method.
    - Cons: `append`/`removeBack` are still somewhat asymmetrical.
4. keep the status quo
    - Pros: Python users are familiar with this interface (as well as Swift users to a certain degree).
    - Cons: This asymmetry might be confusing to users coming from other languages.

**pop(index) method:**

1. deprecate without replacement
    - Pros: Preventing modifications of the middle of the list would make it easier to implement a parallel safe and distributed list. [ working on a separate issue about this topic... ]
    - Cons: although inefficient, this method might be useful for a variety of applications
2. deprecate and replace with `remove`
    - Pros: strong precedent from other languages
    - Cons: this name collides with the existing [`remove`](https://chapel-lang.org/docs/main/modules/standard/List.html#List.list.remove) method. (Java's ArrayList also supports both kinds of remove under the same name)
3. deprecate and replace with `getAndRemove`
   - Pros: the `map` type has a [`getAndRemove`](https://chapel-lang.org/docs/main/modules/standard/Map.html#Map.map.getAndRemove) method. It would be nice to use uniform names for similar functionality across the types.
   - Cons: More verbose than remove. If `pop()` is renamed to something with "remove" in the name, we might want to use something like `getAndRemoveEnd` for symmetry; however, that name is even more verbose.

**push/pop methods on the front of the list:**

1. Don't add them (only support push and pop on the end of the list)
    - Pros: makes it clearer that `list` is more of a vector-list than a linked-list. Users can still use `remove`/`insert` to achieve the same affect.
    - Cons: only supporting these operations on the end of the list makes the API somewhat asymmetrical.
2. add them:
   - Pros: `popFront` is nicer and more index agnostic than `pop(0)` (or `remove(0)`, etc.).
   - Cons: The existence of these methods could imply that the list is implemented as a linked-list, and that pushing/popping on the front of the list is an efficient operation (which it is not). Other languages I looked at only support these methods on their LinkedList types.

**Comment by jeremiah-corrado** [comment]
## Final Design Decision:

During a recent design meeting, a sub-team consisting of myself, @lydia-duncan, and @daviditen, decided to rename `append` and `pop` to `pushBack` and `popBack`.

Old API:
```chapel
var myList = new list();
myList.append(10);
writeln(myList.pop());
```

New API:
```chapel
var myList = new list();
myList.pushBack(10);
writeln(myList.popBack());
```

## Discussion and Rationale:

Pros:
- These names address the concerns noted above about the asymmetry between `append` and `pop`.
- They also address a potential confusion about which end of the list they refer to (particularly for `push` which could refer to either end depending on whether the list is supporting a stack-like or queue-like interface).
- There is a precedent for these names from the C++ vector

Cons:
- There are overloads of `append` that take collections of elements as their arguments. There is not a strong precedent from other languages for "push" methods to accept more than one element. 

We also considered: 

- `push`/`pop`: didn't address the end-ambiguity question
- `append`/`removeLast`: seemed overly wordy and less common in other languages

**Comment by lydia-duncan** [comment]
I believe this was resolved by Jeremiah's #22328, closing
